core/src/mender-client.c

```diff
@@ -1108,186 +1132,338 @@ mender_check_device_compatibility(mender_artifact_ctx_t *mender_artifact_ctx) {
+                                              (const char **)deployment->device_types_compatible,
+                                              deployment->device_types_compatible_size))) {
+        return ret;
     }
 
-#ifdef CONFIG_MENDER_FULL_PARSE_ARTIFACT
 #ifdef CONFIG_MENDER_PROVIDES_DEPENDS
     /* Compare Artifact's depends with the stored provides */
-    if (MENDER_OK != mender_check_device_compatibility(mender_artifact_ctx)) {
-        /* Errors logged by function */
-        mender_client_publish_deployment_status(deployment->id, MENDER_DEPLOYMENT_STATUS_FAILURE);
-        if (mender_client_deployment_needs_set_pending_image) {
-            mender_flash_abort_deployment(mender_client_flash_handle);
-        }
-        goto END;
+    if (MENDER_OK != (ret = mender_check_device_compatibility(mender_artifact_ctx))) {
+        return ret;
     }
+#endif /* CONFIG_MENDER_PROVIDES_DEPENDS */
 
-    /* Add the new provides to the deployment data (we need the artifact context)*/
-    char *new_provides = NULL;
-    if (MENDER_OK != mender_prepare_new_provides(mender_artifact_ctx, &new_provides)) {
-        mender_log_error("Unable to prepare new provides");
-        mender_client_publish_deployment_status(deployment->id, MENDER_DEPLOYMENT_STATUS_FAILURE);
-        if (mender_client_deployment_needs_set_pending_image) {
-            mender_flash_abort_deployment(mender_client_flash_handle);
-        }
-        goto END;
-    }
-    cJSON_AddStringToObject(mender_client_deployment_data, "provides", new_provides);
+    return MENDER_OK;
+}
 #endif /* CONFIG_MENDER_FULL_PARSE_ARTIFACT */
-#endif /* CONFIG_MENDER_PROVIDES_DEPENDS */
 
-    /* Set boot partition */
-    mender_log_info("Download done, installing artifact");
-    mender_client_publish_deployment_status(deployment->id, MENDER_DEPLOYMENT_STATUS_INSTALLING);
-    if (true == mender_client_deployment_needs_set_pending_image) {
-        if (MENDER_OK != (ret = mender_flash_set_pending_image(mender_client_flash_handle))) {
-            mender_log_error("Unable to set boot partition");
-            mender_client_publish_deployment_status(deployment->id, MENDER_DEPLOYMENT_STATUS_FAILURE);
-            goto END;
+static mender_err_t
+mender_client_update_work_function(void) {
+
+    mender_err_t ret = MENDER_OK;
+
+    /* Ensure that the context is initialized to NULL before goto END */
+    mender_artifact_ctx_t *mender_artifact_ctx = NULL;
+
+    /* Check for deployment */
+    mender_api_deployment_data_t *deployment              = calloc(1, sizeof(mender_api_deployment_data_t));
+    char                         *storage_deployment_data = NULL;
+    mender_update_state_t         update_state            = MENDER_UPDATE_STATE_DOWNLOAD;
+    const char                   *deployment_id           = NULL;
+
+    /* reset the currently used update module */
+    mender_client_update_module = NULL;
+
+    {
+        char *artifact_type;
```

Comment:
Initialize `artifact_type` to `NULL`.  Because `mender_storage_get_update_state()` calls `nvs_read_alloc()` that does not guarantee that it is set to `NULL` on failure.